server/models/user-page.js

```diff
@@ -0,0 +1,116 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+'use strict';
+
+// we need a user-pages abstraction so that async scraper jobs
+// can separately handle creating the visit and creating the
+// user page. someday this might also be used by the client-facing API layer.
+
+var underscored = require('underscore.string').underscored;
+var uuid = require('uuid');
+
+var postgres = require('../db/postgres');
+var elasticsearch = require('../db/elasticsearch');
+var log = require('../logger')('server.models.user-page');
+
+var _verbose = function() {
+  var logline = [].join.call(arguments, ', ');
+  log.verbose(logline);
+};
+
+var userPage = {
+  _onFulfilled: function _onFulfilled(msg, callback, results) {
+    _verbose(msg);
+    callback(null, results);
+  },
+  _onRejected: function _onRejected(msg, callback, err) {
+    log.warn(msg);
+    callback(err);
+  },
+  update: function(userId, url, urlHash, title, data, cb) {
+    // fetch the page id, lazily creating it if it doesn't exist,
+    // then update the page with the huge blob of embedly data
+    //
+    // TODO if we don't fire a callback on creation, we should return a promise _or_
+    // fire 'userPage::updated' or 'userPage::updateError' events
+    var name = 'models.user-page.update';
+    _verbose(name + ' called', userId, url);
+    var lazyCreateParams = [uuid.v4(), userId, url, urlHash, title];
+    var lazyCreateUserPageQuery = 'WITH new_page AS (  ' +
+      '  INSERT INTO user_pages (id, user_id, url, raw_url, url_hash, title) ' +
+      '  SELECT $1, $2, $3, $3, $4, $5 ' +
+      '  WHERE NOT EXISTS (SELECT id FROM user_pages WHERE user_id = $2 AND url_hash = $4) ' +
+      '  RETURNING id ' +
+      ') SELECT id FROM new_page ' +
+      'UNION SELECT id FROM user_pages WHERE user_id = $2 AND url_hash = $4';
+    var addExtractedDataQuery = 'UPDATE user_pages SET ( ' +
```

Comment:
It might be nice to have a really simple class that represents the fields and types in a table, to reduce this kind of silly verbosity...oh wait, I'm describing an ORM. It might be nice to use an ORM.